### PR TITLE
Passwords-tool tests

### PR DESCRIPTION
**Pull request by miguelfdez99**
|Related issue|
|---|
|#1640|

## Description

I have added two new files for this.
- **passwords-tool.yml**: In this file, we can see what GitHub Actions do, it is triggered only on pull requests when you modify some file in the directory unattended_installer/passwords_tool/. The first job creates wazuh-install and wazuh-passwords scripts, the second job runs a script and the third job check that errors happen when they should.
- **tests-stack.sh**: This file checks that the password-tool script is working properly.

## Logs example

image

image

image

### CODE REVIEW

File: .github/actions/passwords-tool/script.sh [hunk 1]
```diff
@@ -0,0 +1,91 @@
+#!/bin/bash
```
Comment (alberpilot) [inline .github/actions/passwords-tool/script.sh:1]: Change the file name to `tests-stack.sh`

File: .github/actions/passwords-tool/script.sh [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+users=( admin kibanaserver kibanaro logstash readall snapshotrestore wazuh_admin wazuh_user )
+api_users=( wazuh wazuh-wui )
+
+echo '::group:: Change indexer password, password provided'
+
+bash wazuh-passwords-tool.sh -u admin -p LN*X1v.VNtCZ5sESEtLfijPAd39LXGAI
+ if ! curl -XGET https://localhost:9200/ -u admin:LN*X1v.VNtCZ5sESEtLfijPAd39LXGAI -k -w %{http_code} | grep "200"; then
```
Comment (alberpilot) [inline .github/actions/passwords-tool/script.sh:9]: Why don't use = 401 instead of ! 200?

File: .github/actions/passwords-tool/script.sh [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+  exit 1
+ fi
+echo '::endgroup::'
+
+echo '::group:: Change indexer password, password no provided'
+
+indx_pass="$(bash wazuh-passwords-tool.sh -u admin | awk '/admin/{ print $NF }' | tr -d \' )"
+ if ! curl -XGET https://localhost:9200/ -u admin:"${indx_pass}" -k -w %{http_code} | grep "200"; then
```
Comment (alberpilot) [inline .github/actions/passwords-tool/script.sh:17]: Same

File: .github/actions/passwords-tool/script.sh [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+echo '::group:: Change indexer password, password provided'
+
+bash wazuh-passwords-tool.sh -u admin -p LN*X1v.VNtCZ5sESEtLfijPAd39LXGAI
+ if ! curl -XGET https://localhost:9200/ -u admin:LN*X1v.VNtCZ5sESEtLfijPAd39LXGAI -k -w %{http_code} | grep "200"; then
+  exit 1
+ fi
+echo '::endgroup::'
+
+echo '::group:: Change indexer password, password no provided'
```
Comment (alberpilot) [inline .github/actions/passwords-tool/script.sh:14]:
```suggestion
echo '::group:: Change indexer password without providing it.'
```

File: .github/actions/passwords-tool/script.sh [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+indx_pass="$(bash wazuh-passwords-tool.sh -u admin | awk '/admin/{ print $NF }' | tr -d \' )"
+ if ! curl -XGET https://localhost:9200/ -u admin:"${indx_pass}" -k -w %{http_code} | grep "200"; then
+  exit 1
+ fi
+
+echo '::endgroup::'
+
+echo '::group:: Change all, no API ...'
```
Comment (alberpilot) [inline .github/actions/passwords-tool/script.sh:23]:
```suggestion
echo '::group:: Change all passwords except Wazuh API ones.'
```

File: .github/actions/passwords-tool/script.sh [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+mapfile -t pass < <(bash wazuh-passwords-tool.sh -a | awk '{ print $NF }' | sed \$d | sed '1d' )
+for i in "${!users[@]}"; do
+ if curl -XGET https://localhost:9200/ -u "${users[i]}":"${pass[i]}" -k -w %{http_code} | grep "401"; then
+  exit 1
+ fi
+done
+
+echo '::endgroup::'
+
+echo '::group:: Change all passwords, API credentials ...'
```
Comment (alberpilot) [inline .github/actions/passwords-tool/script.sh:34]:
```suggestion
echo '::group:: Change all passwords.'
```

File: .github/actions/passwords-tool/script.sh [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+for i in "${!api_users[@]}"; do
+ if ! curl -u "${api_users[i]}":"${passapi[i]}" -w "%{http_code}" -k -X GET "https://localhost:55000/security/user/authenticate" | grep "200"; then
+  exit 1
+ fi
+done
+ 
+echo '::endgroup::'
+
+echo '::group:: Change single API user ...'
```
Comment (alberpilot) [inline .github/actions/passwords-tool/script.sh:56]:
```suggestion
echo '::group:: Change single Wazuh API user.'
```

File: .github/actions/passwords-tool/script.sh [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+bash wazuh-passwords-tool.sh -au wazuh -ap "${passapi[0]}" -u wazuh -p BkJt92r*ndzN.CkCYWn?d7i5Z7EaUt63 -A 
+ if ! curl -w "%{http_code}" -u wazuh:BkJt92r*ndzN.CkCYWn?d7i5Z7EaUt63 -k -X GET "https://localhost:55000/security/user/authenticate" | grep "200"; then
+  exit 1
+ fi
+echo '::endgroup::'
+
+echo '::group:: Change all from file, no API ...'
```
Comment (alberpilot) [inline .github/actions/passwords-tool/script.sh:64]:
```suggestion
echo '::group:: Change all passwords except Wazuh API ones using a file.'
```

File: .github/actions/passwords-tool/script.sh [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+
+mapfile -t passfile < <(bash wazuh-passwords-tool.sh -f wazuh-install-files/wazuh-passwords.txt | awk '{ print $NF }' | sed \$d | sed '1d' )
+for i in "${!users[@]}"; do
+ if curl -XGET https://localhost:9200/ -u "${users[i]}":"${passfile[i]}" -k -w %{http_code} | grep "401"; then
+  exit 1
+ fi
+done
+echo '::endgroup::'
+
+echo '::group:: Change all passwords from file using API credentials ...'
```
Comment (alberpilot) [inline .github/actions/passwords-tool/script.sh:74]:
```suggestion
echo '::group:: Change all passwords using a file.'
```

File: .github/workflows/password-tool.yml [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+      - uses: actions/upload-artifact@v3
+        with:
+          name: scripts
+          path: | 
+            unattended_installer/wazuh-install.sh
+            unattended_installer/wazuh-passwords-tool.sh
+          if-no-files-found: error 
+  test-password-tool-1:
+    runs-on: ubuntu-latest
+    needs: Build-password-tool-and-wazuh-install-scripts
```
Comment (alberpilot) [inline .github/workflows/password-tool.yml:26]: Separate jobs. One for Wazuh install, one for test-password-tool-1, and the other for #2.

File: .github/workflows/password-tool.yml [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+        - uses: actions/download-artifact@v3
+          with:
+           name: scripts
+        - name: Install wazuh
+          run: |
+           sudo bash wazuh-install.sh -a
+        - name: Uncompress wazuh install files 
+          run: sudo tar -xvf wazuh-install-files.tar
+        - name: Get wazuh API password
+          run: |
```
Comment (alberpilot) [inline .github/workflows/password-tool.yml:54]: Why those tests are not included in the script?

**Review by alberpilot — CHANGES_REQUESTED** [review]
Please review the requested changes.

File: .github/actions/passwords-tool/script.sh (hunk 1 above)
Comment (miguelfdez99) [reply .github/actions/passwords-tool/script.sh]: Done https://github.com/wazuh/wazuh-packages/pull/1737/commits/054ac37040357937607329818f7f2548254c86cc

File: .github/actions/passwords-tool/script.sh (hunk 2 above)
Comment (miguelfdez99) [reply .github/actions/passwords-tool/script.sh]: Done https://github.com/wazuh/wazuh-packages/pull/1737/commits/054ac37040357937607329818f7f2548254c86cc

File: .github/actions/passwords-tool/script.sh [hunk 12]
```diff
@@ -0,0 +1,91 @@
+#!/bin/bash
+
+users=( admin kibanaserver kibanaro logstash readall snapshotrestore wazuh_admin wazuh_user )
+api_users=( wazuh wazuh-wui )
+
+echo '::group:: Change indexer password, password provided'
```
[the remark this answers is not in the document]
Comment (miguelfdez99) [reply .github/actions/passwords-tool/script.sh]: Done https://github.com/wazuh/wazuh-packages/pull/1737/commits/054ac37040357937607329818f7f2548254c86cc

File: .github/actions/passwords-tool/script.sh (hunk 3 above)
Comment (miguelfdez99) [reply .github/actions/passwords-tool/script.sh]: Done https://github.com/wazuh/wazuh-packages/pull/1737/commits/054ac37040357937607329818f7f2548254c86cc

File: .github/actions/passwords-tool/script.sh (hunk 4 above)
Comment (miguelfdez99) [reply .github/actions/passwords-tool/script.sh]: Done https://github.com/wazuh/wazuh-packages/pull/1737/commits/054ac37040357937607329818f7f2548254c86cc

File: .github/actions/passwords-tool/script.sh (hunk 6 above)
Comment (miguelfdez99) [reply .github/actions/passwords-tool/script.sh]: Done https://github.com/wazuh/wazuh-packages/pull/1737/commits/054ac37040357937607329818f7f2548254c86cc

File: .github/actions/passwords-tool/script.sh (hunk 7 above)
Comment (miguelfdez99) [reply .github/actions/passwords-tool/script.sh]: Done https://github.com/wazuh/wazuh-packages/pull/1737/commits/054ac37040357937607329818f7f2548254c86cc

File: .github/actions/passwords-tool/script.sh (hunk 9 above)
Comment (miguelfdez99) [reply .github/actions/passwords-tool/script.sh]: Done https://github.com/wazuh/wazuh-packages/pull/1737/commits/054ac37040357937607329818f7f2548254c86cc

File: .github/workflows/password-tool.yml (hunk 11 above)
Comment (miguelfdez99) [reply .github/workflows/password-tool.yml:54]: Those tests are checking expected failures, the script is checking that everything succeeds.

**Review by okynos — APPROVED** [review]

**Review by DFolchA — APPROVED** [review]
LGTM

**Pull request merged**